ably/error.go

```diff
@@ -0,0 +1,263 @@
+package ably
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+
+	"github.com/ably/ably-go/ably/proto"
+)
+
+func tostatus(code int) int {
+	switch status := code / 100; status {
+	case 400, 401, 403, 404, 405, 500:
+		return status
+	default:
+		return 0
+	}
+}
+
+// Error describes error returned from Ably API. It always has non-zero error
+// code. It may contain underlying error value which caused the failure
+// condition.
+type Error struct {
+	Code   int   // internal error code
+	Status int   // HTTP status code
+	Err    error // underlying error responsible for the failure; may be nil
+}
+
+// Error implements builtin error interface.
+func (err *Error) Error() string {
+	if err.Err != nil {
+		return err.Err.Error()
+	}
+	return errCodeText[err.Code]
+}
+
+func newError(code int, err error) error {
+	switch err := err.(type) {
+	case *Error:
+		return err
+	case net.Error:
+		if err.Timeout() {
+			return &Error{Code: ErrCodeTimeout, Status: 500, Err: err}
+		}
+	}
+	return &Error{
+		Code:   code,
+		Status: tostatus(code),
+		Err:    err,
+	}
+}
+
+func newErrorf(code int, format string, v ...interface{}) error {
+	return &Error{
+		Code:   code,
+		Status: tostatus(code),
+		Err:    fmt.Errorf(format, v...),
+	}
+}
+
+func newErrorProto(err *proto.Error) error {
+	if err == nil {
+		return nil
+	}
+	return &Error{
+		Code:   err.Code,
+		Status: err.StatusCode,
+		Err:    errors.New(err.Message),
+	}
+}
+
+func checkError(resp *http.Response) error {
+	const maxContentLength = 32 * 1024 // 32 KiB
+	if resp.StatusCode/100 < 3 {
+		return nil
+	}
+	n := resp.ContentLength
+	if n <= 0 || n >= maxContentLength {
+		n = maxContentLength
+	}
+	body := &proto.Error{}
+	if e := json.NewDecoder(io.LimitReader(resp.Body, n)).Decode(body); e != nil {
```

Comment:
I feel like we are asking for trouble here. Why not just do a `ioutil.ReadAll(resp.Body)` here, instead of expecting the server to return the right content length?